ddtrace/tracer/textmap.go

```diff
@@ -262,53 +262,74 @@ func (p *chainedPropagator) Inject(spanCtx ddtrace.SpanContext, carrier interfac
+// a previous propagator has succeeded so long as the trace-ids match.
+// Furthermore, if we have already successfully extracted a trace context and a
+// subsequent trace context has conflicting trace information, such information will
+// be relayed in the returned SpanContext with a SpanLink.
 func (p *chainedPropagator) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
 	var ctx ddtrace.SpanContext
+	var links []ddtrace.SpanLink
 	for _, v := range p.extractors {
-		if ctx != nil {
-			// A local trace context has already been extracted.
-			pw3c, isW3C := v.(*propagatorW3c)
-			if !isW3C {
-				continue // Ignore other propagators.
+		firstExtract := (ctx == nil) // ctx stores the most recently extracted ctx across iterations; if it's nil, no extractor has run yet
+		extractedCtx, err := v.Extract(carrier)
+		if firstExtract {
+			if err != nil && err != ErrSpanContextNotFound { // We only care if the first extraction returns an error because this breaks distributed tracing
+				return nil, err
 			}
-			w3cCtx, err := pw3c.Extract(carrier)
-			if err == nil && w3cCtx.(*spanContext).TraceID128() == ctx.(*spanContext).TraceID128() {
-				pw3c.propagateTracestate(ctx.(*spanContext), w3cCtx.(*spanContext))
-				if w3cCtx.SpanID() != ctx.SpanID() {
-					var ddCtx *spanContext
-					if ddp := getDatadogPropagator(p); ddp != nil {
-						if ddSpanCtx, err := ddp.Extract(carrier); err == nil {
-							ddCtx, _ = ddSpanCtx.(*spanContext)
+			if p.onlyExtractFirst { // Return early if only performing one extraction
+				return extractedCtx.(*spanContext), nil
+			}
+			ctx = extractedCtx
+		} else { // A local trace context has already been extracted
+			extractedCtx2, ok1 := extractedCtx.(*spanContext)
+			ctx2, ok2 := ctx.(*spanContext)
+			// If we can't cast to SpanContextW3C, we can't propgate tracestate or create span links
+			if !ok1 || !ok2 {
+				continue
+			}
+			if extractedCtx2.TraceID128() == ctx2.TraceID128() {
+				if pW3C, ok := v.(*propagatorW3c); ok {
+					pW3C.propagateTracestate(ctx2, extractedCtx2)
+					// If trace IDs match but span IDs do not, use extractedCtxW3c span ID for parenting
+					if extractedCtx2.SpanID() != ctx2.SpanID() {
+						var ddCtx *spanContext
+						// Grab the datadog-propagated spancontext again
+						if ddp := getDatadogPropagator(p); ddp != nil {
+							if ddSpanCtx, err := ddp.Extract(carrier); err == nil {
+								ddCtx, _ = ddSpanCtx.(*spanContext)
+							}
 						}
+						overrideDatadogParentID(ctx.(*spanContext), extractedCtx.(*spanContext), ddCtx)
 					}
-					overrideDatadogParentID(ctx.(*spanContext), w3cCtx.(*spanContext), ddCtx)
 				}
+			} else { // Trace IDs do not match - create span links
+				link := ddtrace.SpanLink{TraceID: extractedCtx2.TraceID(), SpanID: extractedCtx2.SpanID(), TraceIDHigh: extractedCtx2.TraceIDUpper(), Attributes: map[string]string{"reason": "terminated_context", "context_headers": "tracecontext"}}
```

Comment:
@mtoffl01 FYI the `context_headers` field is supposed to hold whichever propagator type was used to extract the violating context. For instance, if the propagator type was `Datadog` and we have violating trace IDs, we want `context_headers` to have the value of `datadog`